docs/rfcs/xxxx-river.md

```diff
@@ -0,0 +1,404 @@
+final config state should be.
+
+### Why not HCL?
+
+> For some background, it's important to note that HCL can be considered two
+> separate projects: hashicorp/hcl (the language and expression evaluator) and
+> zclconf/go-cty (the value and type system used by HCL).
+
+HCL was the obvious first choice for the Flow prototype: it supports
+expressions, you can expose functions for users to call, and its syntax has a
+small learning curve.
+
+However, I found the [schema-driven processing][] API exposed by HCL to be
+difficult to work with for Flow, requiring a lot of boilerplate. While there is
+a library to interoperate with tagged Go structs, it was insufficient for
+passing around arbitrary Go values, requiring me to [fork][gohcl] both
+github.com/hashicorp/hcl/v2/gohcl and github.com/zclconf/go-cty/cty/gocty to
+reduce boilerplate. This fork contains a non-trivial amount of changes that
+would need to be contributed upstream to be tenable long-term.
+
+Additionally, there is desired functionality that is not supported today in
+HCL/go-cty:
+
+1. A stronger focus on performance and memory usage, changing go-cty to operate
+   around Go values instead of converting Go values to a custom representation.
+2. Ability to disable go-cty's requirement that strings are UTF-8 encoded
+3. Pass around functions as go-cty values (e.g., to allow a clustering
+   component to expose a function to check for ownership of key against a hash
+   ring)
+4. Ability to declare local variables in a scope without needing a `locals`
+   block like as seen in Terraform.
+
+The combination of desired changes across gohcl and go-cty, the fork that was
+already necessary to make it easier to adopt HCL for Flow, and the desire to
+have a stronger interaction with arbitrary Go values led to the decision that a
+new Flow-specific language was warranted.
+
+### Why now?
+
+Grafana Agent Flow is already a dramatic change to the Agent. To avoid users
+being exhausted from the frequency of dramatic changes, it would be ideal for
+Grafana Agent Flow to ship with River instead of eventually migrating to River.
+
+## Minimizing impact
+
+New languages always have some amount of learning curve, and if the learning
+curve is too steep, the language will fail to be adopted.
+
+We will minimize this impact of a new language by:
+
+* Minimizing the learning curve as much as possible by not creating
+  too many novel ideas at the language level.
+
+* Tend the syntax towards allowing users to copy-and-paste examples to learn as
+  they go.
+
+* Heavily document the language so that all questions a user may have is
+  answered.
+
+* Ensuring that error messages explain the problem and the resolution is
+  obvious.
+
+## Proposal
+
+River's syntax is inspired by HCL. However, some of the syntax will be changed
+from HCL to make River more easily identifiable as a different language and
+avoid a situations where users confuse the two.
```

Comment:
Not related to the RFC, but we might want to balance new changes with being able to use existing HCL tooling (eg. syntax highlighting)

Reply:
Yeah, feel free to disregard this comment. When I saw how quickly you were able to produce VSCode/Vim plugins, it makes less sense to worry about tooling, and as you said gives us absolute freedom to choose our path.